File: src/domain/models/sync.ts
export type SyncEntityType =
  | 'task'
  | 'task_reminder'
  | 'tag'
  | 'attachment'
  | 'habit'
  | 'habit_log'
  | 'setting';

export type SyncOperationType = 'create' | 'update' | 'reorder' | 'delete';

export type SyncOperationStatus = 'pending' | 'syncing' | 'acked' | 'failed';

export type SyncEntityStatus =
  | 'clean'
  | 'pending'
  | 'syncing'
  | 'failed'
  | 'conflicted'
  | 'deleted';

export interface SyncOperation {
  op_id: string;
  entity_type: SyncEntityType;
  entity_id: string;
  operation: SyncOperationType;
  base_revision: number | null;
  payload: unknown;
  client_time: string;
  device_id: string;
  status: SyncOperationStatus;
  retry_count: number;
  last_error: string | null;
  created_at: string;
}

export interface CreateSyncOperationInput {
  op_id?: string;
  entity_type: SyncEntityType;
  entity_id: string;
  operation: SyncOperationType;
  base_revision?: number | null;
  payload: unknown;
  client_time?: string;
  device_id: string;
}

export interface SyncConflict {
  id: string;
  entity_type: SyncEntityType;
  entity_id: string;
  local_payload: unknown;
  remote_payload: unknown;
  created_at: string;
  resolved_at: string | null;
}

export interface CreateSyncConflictInput {
  id?: string;
  entity_type: SyncEntityType;
  entity_id: string;
  local_payload: unknown;
  remote_payload: unknown;
  created_at?: string;
}

export interface SyncMetaEntry {
  key: string;
  value: string;
}

File: src-tauri/src/models/sync.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SyncMetaEntry {
    pub key: String,
    pub value: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SyncOperation {
    pub op_id: String,
    pub entity_type: String,
    pub entity_id: String,
    pub operation: String,
    pub base_revision: Option<i64>,
    pub payload: Value,
    pub client_time: String,
    pub device_id: String,
    pub status: String,
    pub retry_count: i32,
    pub last_error: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CreateSyncOperationRequest {
    pub op_id: Option<String>,
    pub entity_type: String,
    pub entity_id: String,
    pub operation: String,
    pub base_revision: Option<i64>,
    pub payload: Value,
    pub client_time: Option<String>,
    pub device_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SyncConflict {
    pub id: String,
    pub entity_type: String,
    pub entity_id: String,
    pub local_payload: Value,
    pub remote_payload: Value,
    pub created_at: String,
    pub resolved_at: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CreateSyncConflictRequest {
    pub id: Option<String>,
    pub entity_type: String,
    pub entity_id: String,
    pub local_payload: Value,
    pub remote_payload: Value,
    pub created_at: Option<String>,
}
